### PR TITLE
Add cache for Elderly and Volunteers

### LINKED ISSUE #50: Add cache implementation for Elderly and Volunteer
- Extract `UniquePersonList` that takes in a generic `T` which extends from `Person`
- Update storage implementation to serialize Elderly and Volunteers list as well
- Update necessary test cases

### PR DESCRIPTION
**Pull request by Zxun2**
Fixes #50 

Updates: 
- Added storage model for Elderly and Volunteer.
- Updated the storage mechanism for Volunteers and Elderly to be stored separately. In total, we should have 3 data files, one for Volunteer, one for Elderly and one more for any application wide settings -- this also currently used to store pairs (not user prefs!). @gohyongjing maybe you want to separate it out to a file as well? 
- Autosave for elderly and volunteers is **not implemented** in this pr as it is getting quite big. 
- Test cases need to be relooked at, to accommodate the above changes. (I will add more testcases in another PR since this PR is getting quite huge for some reason)
- Clean up files with `address book`
- More abstraction in general

Tests passed 328/328

### CODE REVIEW

**Comment by Zxun2** [comment]
@AY2223S2-CS2103T-W12-1/developers hey guys, do you think it's better if we store all data in 1 file, `friendlylink.json` or separate them out to their respective jsons, e.g. `volunteers.json` and `elderly.json`?

File: src/main/java/seedu/address/ui/UiManager.java
Comment (wz2k) [inline src/main/java/seedu/address/ui/UiManager.java:55]: Very nice 👍  but to follow the rest of the code, requireNonNull() can be imported.

**Review by wz2k — COMMENTED** [review]
LGTM. 👍

**Comment by wz2k** [comment]
> @AY2223S2-CS2103T-W12-1/developers hey guys, do you think it's better if we store all data in 1 file, `friendlylink.json` or separate them out to their respective jsons, e.g. `volunteers.json` and `elderly.json`?

I feel separate files are better but it really depends on the complexity of the changes that needs to be made to the current code base and the test cases needed to ensure our code stays bug free.

**Comment by zhacatomn** [comment]
Looks good, although there'll be some conflicts between this and #49. In #49, I added `ModelStub` as a static class under the `CommandTestUtil` class, while here it is in its own separate class: which would you prefer? Additionally, the `Person` needs to be replaced with `Volunteer` in the relevant areas. 

Thanks for the good work!

**Comment by heeeyi** [comment]
I feel it is better indeed better to extract out `ModelStub` separately, like it is in your fork. I have just pulled request the #51, which still use the old one where `ModelStub` was in separate classes. Will update to the new version once your fork is merged.

**Comment by zhacatomn** [comment]
> I feel it is better indeed better to extract out `ModelStub` separately, like it is in your fork. I have just pulled request the #51, which still use the old one where `ModelStub` was in separate classes. Will update to the new version once your fork is merged.

Okay, here's what I'll do: for #49, I will revert my changes to `ModelStub`, and just add it as a static class to `AddVolunteerCommand`. After I merge in the PR, @Zxun2 could you settle the conflicts? Sorry for the trouble.

**Comment by zhacatomn** [comment]
> @AY2223S2-CS2103T-W12-1/developers hey guys, do you think it's better if we store all data in 1 file, `friendlylink.json` or separate them out to their respective jsons, e.g. `volunteers.json` and `elderly.json`?

I think separating will be better: Basically treating each json file as its own "DB".

**Comment by Zxun2** [comment]
> > I feel it is better indeed better to extract out `ModelStub` separately, like it is in your fork. I have just pulled request the #51, which still use the old one where `ModelStub` was in separate classes. Will update to the new version once your fork is merged.
> 
> Okay, here's what I'll do: for #49, I will revert my changes to `ModelStub`, and just add it as a static class to `AddVolunteerCommand`. After I merge in the PR, @Zxun2 could you settle the conflicts? Sorry for the trouble.

@zhacatomn Sure!

**Comment by Zxun2** [comment]
@gohyongjing please have a look, and **approve** the pull request when you're done. let me know if there's any changes you want as well.

File: src/main/java/seedu/address/storage/elderly/JsonSerializableElderly.java
Comment (gohyongjing) [inline src/main/java/seedu/address/storage/elderly/JsonSerializableElderly.java:21]: Perhaps we should change these to plural to to make it  clear that this class is a list of elderlies
```suggestion
@JsonRootName(value = "elderlies")
public class JsonSerializableElderlies {
```

We can change this in the next pull request

File: src/main/java/seedu/address/storage/volunteer/JsonSerializableVolunteer.java
Comment (gohyongjing) [inline src/main/java/seedu/address/storage/volunteer/JsonSerializableVolunteer.java:22]: This constant seems to be repeated accrosee multiple classes, perhaps change it to be more specific?
```suggestion
    public static final String MESSAGE_DUPLICATE_VOLUNTEER = "Persons list contains duplicate volunteer(s).";
```

File: src/test/data/JsonAddressBookStorageTest/invalidPersonAddressBook.json
Comment (gohyongjing) [inline src/test/data/JsonAddressBookStorageTest/invalidPersonAddressBook.json:9]: Are we not putting these in a separate json file?

File: src/test/java/seedu/address/logic/commands/ModelStub.java
Comment (gohyongjing) [inline src/test/java/seedu/address/logic/commands/ModelStub.java:19]: I like that model stub is abstracted out, saves many lines of code 👍

**Review by gohyongjing — COMMENTED** [review]
Looks good! Will update my code to match yours after this PR is merged

**Review by gohyongjing — APPROVED** [review]
LGTM! do make the changes above or let me know to put them in a separate issue if they are too large to be resolved in this PR

File: src/test/data/JsonAddressBookStorageTest/invalidPersonAddressBook.json
Comment (Zxun2) [reply src/test/data/JsonAddressBookStorageTest/invalidPersonAddressBook.json:9]: Yup my bad

File: src/main/java/seedu/address/storage/JsonSerializableFriendlyLink.java
Comment (Zxun2) [reply src/main/java/seedu/address/storage/JsonSerializableFriendlyLink.java:60]: Yup agreed. I'll do this in another pr when i abstract the `JsonSerializableEntity` class. Thanks for spotting this!

**Comment by Zxun2** [comment]
@gohyongjing merge when you're ready!

File: src/main/java/seedu/address/storage/volunteer/JsonSerializableVolunteer.java
Comment (Zxun2) [reply src/main/java/seedu/address/storage/volunteer/JsonSerializableVolunteer.java]: Yup my bad

**Pull request merged**
